core/src/main/java/org/apache/iceberg/NameMappingUpdate.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.iceberg.mapping.MappedField;
+import org.apache.iceberg.mapping.MappedFields;
+import org.apache.iceberg.mapping.NameMapping;
+import org.apache.iceberg.mapping.NameMappingParser;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.types.Types;
+
+/**
+ * NameMapping evolution API implementation.
+ */
+class NameMappingUpdate implements UpdateNameMapping {
+  private final Schema schema;
+  private final NameMapping nameMapping;
+  private final UpdateProperties updateProperties;
+  private final Map<Integer, Iterable<String>> adds = new ConcurrentHashMap<>();
+
+  NameMappingUpdate(TableOperations ops) {
+    TableMetadata tableMetadata = ops.current();
+    this.schema = tableMetadata.schema();
+    this.nameMapping = NameMappingParser.fromJson(tableMetadata.properties().get(TableProperties.DEFAULT_NAME_MAPPING));
+    this.updateProperties = new PropertiesUpdate(ops);
+  }
+
+  /**
+   * For testing only.
+   */
+  NameMappingUpdate(
+      Schema schema, NameMapping nameMapping, UpdateProperties updateProperties) {
+    this.schema = schema;
+    this.nameMapping = nameMapping;
+    this.updateProperties = updateProperties;
+  }
+
+  private void addAliases(int fieldId, Iterable<String> aliases) {
+    adds.compute(fieldId, (dontCare, oldValue) -> {
+      if (oldValue == null) {
+        return aliases;
+      } else {
+        return ImmutableList.<String>builder().addAll(aliases).build();
+      }
+    });
+  }
+
+  @Override
+  public UpdateNameMapping addAliases(String name, Iterable<String> aliases) {
+    Preconditions.checkArgument(schema.findField(name) != null, "Cannot find column: %s", name);
+    int fieldId = schema.findField(name).fieldId();
```

Comment:
I don't think we should require a schema lookup here. When adding a field to the mapping, we can add the ID from the schema, but we shouldn't require a field in the schema to update the mapping. There's no need to tie the two together.